### DISCUSSION TITLE
How do I enable Zsh completions?

### DISCUSSION THREAD

**Discussion by mattmc3** [mattmc3/antidote#74] [Q&A] [answered]
I have included [zsh-users/zsh-completions](https://github.com/zsh-users/zsh-completions) in my `.zsh_plugins.txt`, but I still get no completions. How do I make completions work properly with antidote?

**Comment by mattmc3** [comment, marked as answer]
Antidote is a plugin manager, not a complete Zsh framework. If you have used [Oh-My-Zsh](https://github.com/ohmyzsh/ohmyzsh) or [Prezto](https://github.com/sorin-ionescu/prezto) in the past, you might be used to a ton of out-of-the-box features like TAB completions. zsh-users/zsh-completions may seem like it does the job, but it still requires you to run `compinit`. To get completions working properly with an antidote-based Zsh config, you’ll need to run `compinit` yourself once you've finished building your `fpath` with the location of your `_completions` files, or simply use a completions plugin. The good news it that’s easy to do!

If you want to manually enable Zsh completions, add the following snippet to your .zshrc after you’ve loaded your plugins.

```zsh
# .zshrc
autoload -Uz compinit && compinit
```

Or, you can also use a completion plugin. Some good ones are [belak/zsh-utils](https://github.com/belak/zsh-utils) and [mattmc3/zephyr](https://github.com/mattmc3/zephyr).

```zsh
# .zsh_plugins.txt

# add more completions with plugins
zsh-users/zsh-completions

# now, use a plugin that runs compinit for you to initialize your completions
mattmc3/zephyr path:completion
```

**Comment by fdw** [comment]
According to your docs, I need to call `autoload -Uz compinit && compinit` after I've sourced the plugins. However, some of my plugins (https://github.com/gradle/gradle-completion, https://github.com/lukechilds/zsh-better-npm-completion) then throw an error `command not found: compdef`. Do I also need to call compinit before sourcing the plugins? That sounds weird, too...

**Reply by mattmc3** [reply]
In your case, where you have plugins that use `#compdef`, you may find it helpful to use a plugin that initializes your completions before those plugins per the second part of my answer above:

```zsh
... # other plugins

# add plugins that use fpath for their completions BEFORE calling compinit
zsh-users/zsh-completions

# now, use a plugin that runs compinit for you to initialize your completions - there are a few
# mattmc3/zephyr path:completion
# zshzoo/compinit
belak/zsh-utils path:completion

# AFTER calling compinit, load completion plugins that use completion features directly (compdef)
gradle/gradle-completion
lukechilds/zsh-better-npm-completion

# You can still load other plugins after compinit, but be aware that any additional fpath modifications
# won't get picked up by the completion system now, and the compdump file will need rebuilt
zsh-users/zsh-history-substring-search
...

```

You can read more about the [Zsh completion system here](https://zsh.sourceforge.io/Doc/Release/Completion-System.html).
